Add plumbing for heighmap-aware shroud and map bounds checks.

pchote opened the pull request:
This introduces a new coordinate system ("projected map cells", or `PPos`) for code that deals with the map border and shroud.  For mods that don't use terrain height `PPos` maps identically to `MPos`.

From the first commit message:

```
PPos is best thought of as a cell grid applied in
screen space.  Multiple cells with different
terrain heights may be projected to the same PPos,
or to multiple PPos if they do not align with the
screen grid (i.e. odd-height or sloped).

PPos coordinates are used primarily for map edge
checks and shroud / visibility queries.
```

For the sake of keeping the PR as reviewable as possible I have limited this to what we need for the shroud.  Future PRs will deal with the map-edge coordinate helpers which are fiddly and harder to review.

Expected behaviour with this PR applied is for there to be no visible changes in TD/RA/D2K, and for the TS shroud rendering to behave logically.  Edge of map behaviour in TS should be no worse than in current bleed.

pchote committed:
Convert shroud calculations and rendering to PPos.

pchote committed:
Introduce initial PPos plumbing.

PPos is best thought of as a cell grid applied in
screen space.  Multiple cells with different
terrain heights may be projected to the same PPos,
or to multiple PPos if they do not align with the
screen grid.

PPos coordinates are used primarily for map edge
checks and shroud / visibility queries.

pchote commented:
I won't try and force it, but I think we should aim to get this in for the next release.  We now have several community mods building off TS, and this would be a big benefit to them.

pchote commented:
I decided to include the proper projection code, as "shroud versus map edge helpers" is a better split for testing than "plumbing versus fiddly calculations".  Updated the original description to match.

MicroJOo commented:
exception.log
http://pastebin.com/td0XV5z6
graphics.log
http://pastebin.com/Tf51kKyP

pchote commented:
Fixed the above crash, and we gave this a stress-test with an awesome 2v2 match (which ended prematurely thanks to #8789).

Replay is available at https://www.dropbox.com/s/sli7p8iec6d5btc/OpenRA-pr8781-33e6f4e.orarep?dl=0 if anyone wants to hunt for bugs in it.

pchote commented:
> Expected behaviour with this PR applied is for there to be no visible changes in TD/RA/D2K.

This isn't quite true: aircraft visibility in these mods is now centered on the aircraft itself, instead of the ground below it.  When we originally discussed this change (months ago) this was considered a good thing.

GraionDilach commented:
>aircraft visibility in these mods is now centered on the aircraft itself, instead of the ground below it

You're a hero.

penev92 commented:
Just putting it out there - that aircraft visibility sucks.

GraionDilach commented:
It always did. I guess we can amend the game later anyway.

To me, that falls into the same basket as the Mammoth MK2 having a working Burst: 2 on it's railgun/ Devils Tongue has a fixed rate of fire on the flamethrower - while yes, according to the INI, these were intended, but due to how the engine works, any unit which fires a particle-based weapon in TS/RA2 (railgun/flamesare) cannot refire that weapon until the previous instance still has particles of it (railgun spiral effect, fire streams), leading to DT having a slower rate of fire on bigger ranges and MMkII railgun never bursting.

RoosterDragon commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -695,6 +785,34 @@ public CPos CellContaining(WPos pos)
+			return cellProjection[uv];
+		}
+
+		public MPos[] Unproject(PPos puv)
```
This is currently unused, will you be needing it later?

pchote replied:
Yeah, this will be used in future PRs that resolve the `// TODO: Account for terrain height` through Map.

RoosterDragon commented on OpenRA.Game/Map/ProjectedCellRegion.cs:
```diff
@@ -0,0 +1,195 @@
+			public void Dispose() { }
+		}
+
+		public struct MapCoordsRegion : IEnumerable<MPos>
```
We should probably extract a common version of this.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -262,25 +270,26 @@ public bool IsExplored(MPos uv)
 			if (!map.Contains(uv))
 				return false;
 
+			return map.ProjectToShroudPlane(uv).Any(IsExplored);
```
Cache this delegate please.

RoosterDragon commented on OpenRA.Game/Traits/World/Shroud.cs:
```diff
@@ -294,39 +303,40 @@ bool IsExploredCore(MPos uv)
+			if (!visibleCount.Contains(uv))
 				return false;
 
+			return map.ProjectToShroudPlane(uv).Any(IsVisible);
```
Cache this delegate please.

RoosterDragon commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -652,7 +737,12 @@ public bool Contains(CPos cell)
 		public bool Contains(MPos uv)
 		{
-			return Bounds.Contains(uv.U, uv.V);
+			return Bounds.Contains(uv.U, uv.V) && ProjectedCellsCovering(uv).All(Contains);
```
Cache this delegate or de-LINQ this please.

RoosterDragon commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -428,6 +430,85 @@ void PostInit()
+		{
+			var mapHeight = MapHeight.Value;
+			if (!mapHeight.Contains(uv))
+				yield break;
```
Avoid the compiler generated enumerable and return Enumerable.Empty<PPos>() on this fast path.

RoosterDragon commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -428,6 +430,85 @@ void PostInit()
+			if (height == 0)
+			{
+				yield return (PPos)uv;
+				yield break;
```
Avoid the compiler generated enumerable and return a new array on this fast path.

The rest of the method can retain the yield statements, but this improves memory allocations on flat maps significantly.

RoosterDragon commented:
Played a couple of test games on TS and RA and the shroud seems to work just fine on both. Awesome. Overall it seems reasonably performant too.

I think we might do well to add some much explanatory descriptions about the difference between MPos/CPos/PPos/WPos and their regions though - we've got quite a few co-ordinates systems declared in code and if you're not already familiar with them they're pretty scary.

pchote commented:
Rebased, cached delegates, and fixed a major derp where the shroud was manually projecting everything instead of using the cached mapping.  These changes improve performance without affecting the ingame logic - the replay above still plays without desyncing.

pchote committed:
Extract MapCoordsRegion from CellRegion.

pchote commented:
Updated to remove `MapCoordsRegion` duplication.

pchote commented:
Fixed.

Mailaender commented:
Works as promised: :white_check_mark:

pchote commented:
Updated to add a Contains check to Map.Unproject and correct the bounds check in Map.Contains(MPos).

pchote committed:
Temporarily work around RadarWidget corruption.

pchote commented:
Turns out that correcting the bounds check in Map.Contains exposes massive problems in RadarWidget.  I've reverted that change (so it is now back to the same code that was :+1:'d above), but left it in its own commit for future revertability and wrote a descriptive comment.

obrakmann commented:
Something is wrong when you're trying to place the advanced power plant, it looks like this now, was ok before. Other buildings were ok, though, so I guess it's rather a bug in the setup of the actor than with this code.

![screenshot from 2015-07-28 00 06 13](https://cloud.githubusercontent.com/assets/4331210/8919155/6ef8e4ee-34bd-11e5-9e1b-e6d2c002c8dc.png)

obrakmann commented:
Also, in the editor it looks like this now. I suppose that is expected? (Wouldn't know what else to do there when showing the area inside the cordon. Probably show it fogged, though. Either way, not sure if that should be part of this PR.)
![screenshot from 2015-07-28 00 16 05](https://cloud.githubusercontent.com/assets/4331210/8919264/405adf7e-34be-11e5-86e1-bbad6c2bc6cc.png)

Phrohdoh commented:
I tested this on RA2 and noticed no difference at all, positive or negative.
I'm not sure if that is good.

pchote commented:
> Also, in the editor it looks like this now. I suppose that is expected? 

Yes.  This is a consequence of our fixed-size `CellLayer`, and we would need a completely new layer type just for shroud to fix this cosmetic editor issue.

pchote commented:
> Something is wrong when you're trying to place the advanced power plant, it looks like this now, was ok before.

I can reproduce this on bleed, so its not introduced by this pr.  It only happens on "A River Runs Near It".

obrakmann commented:
Shroud now works in ts, and continues to work in the regular mods. Code seems fine as far as I can be a judge of that. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/bce1ad43726735b35379e14fe47900febac9fb00%5E...bce1ad43726735b35379e14fe47900febac9fb0)
